### PR TITLE
fix: list log files with checkpoint when version is None

### PR DESCRIPTION
**Pull request by mervynzhang**
Previously, if read_last_checkpoint returned a valid _last_checkpoint hint but the version requested was None (i.e. requesting the lastest version) we would still list_log_files without listing from the checkpoint hint. This change will detect when we have a valid _last_checkpoint and request the latest version and instead list_log_files_with_checkpoint

A side effect, we will get following error when load a table via s3proxy with a lot log files (more than 1000), but the same issue doesn't occur on delta-rs, after some troubleshooting, the error can be fixed by add this line from delta-rs

https://github.com/delta-io/delta-rs/blob/main/crates/core/src/kernel/snapshot/log_segment.rs#L98

```
ObjectStore(
    Generic {
        store: "S3",
        source: ListRequest {
            source: Client {
                status: 400,
                body: Some(
                    "<?xml version='1.0' encoding='UTF-8'?><Error><Code>InvalidArgument</Code><Message>Bad Request</Message><RequestId>4442587FB7D0A2F9</RequestId></Error>",
                ),
            },
        },
    },
)

### CODE REVIEW

**Review by zachschuermann — COMMENTED** [review]
Hi @mervynhang thanks for the PR! It does look like we aren't reading from the last checkpoint when asking for the latest version - thanks for fixing. Curious though if there is some constraint on the LIST API that's causing the 400? we should still be able to read the table without the checkpoint

**Comment by mervynzhang** [comment]
Hi @zachschuermann ,

We have to use s3proxy in your environment,  s3proxy require that continuation-token and start-after cannot be both set

https://github.com/gaul/s3proxy/blob/4976e170c3c068275201a79d329f953f685fb808/src/main/java/org/gaul/s3proxy/S3ProxyHandler.java#L1389C17-L1389C35

delta-kernel-rs will set start-after
https://github.com/delta-incubator/delta-kernel-rs/blob/main/kernel/src/snapshot.rs#L367

but when there are more than 1000 items, continuation-token will be set automatically and s3proxy will return error. 
https://github.com/apache/arrow-rs/blob/3e02689e3464bc8cf929a0d116888fb6f59999fa/object_store/src/aws/client.rs#L726

**Comment by zachschuermann** [comment]
Ah great, thanks for the context. So, put another way, attempting to list > 1000 items isn't supported for `ObjectStore` on S3 with s3proxy? Is that a known issue?

**Comment by zachschuermann** [comment]
(the fix here is a useful optimization but more of a workaround for the underlying issue in ObjectStore/S3proxy it seems)

File: kernel/src/snapshot.rs [hunk 1]
```diff
@@ -149,6 +149,9 @@ impl Snapshot {
         // List relevant files from log
         let (mut commit_files, checkpoint_files) =
             match (read_last_checkpoint(fs_client.as_ref(), &log_url)?, version) {
+                (Some(cp), None) => {
+                    list_log_files_with_checkpoint(&cp, fs_client.as_ref(), &log_url)?
+                },
```
Comment (zachschuermann) [inline kernel/src/snapshot.rs:154]:
```suggestion
                (Some(cp), None) => {
                    list_log_files_with_checkpoint(&cp, fs_client.as_ref(), &log_url)?
                }
```

**Review by zachschuermann — CHANGES_REQUESTED** [review]
Hi @mervynhang if we modify the golden_tables tests (in my comment above) we should be good to go!

**Comment by mervynzhang** [comment]
> Hi @mervynhang if we modify the golden_tables tests (in my comment above) we should be good to go!

Hi @zachschuermann  golden_tables tests has been updated, please let me know if anything else is needed. Thanks

**Comment by zachschuermann** [comment]
could we update the PR description to reflect the changes being made? feel free to keep the original context but add something like:

previously, if `read_last_checkpoint` returned a valid `_last_checkpoint` hint but the version requested was `None` (i.e. requesting the lastest version) we would still `list_log_files` without listing from the checkpoint hint. This change will detect when we have a valid `_last_checkpoint` and request the latest version and instead `list_log_files_with_checkpoint`

**Comment by mervynzhang** [comment]
Hi @zachschuermann , PR description updated. Thanks

**Review by zachschuermann — APPROVED** [review]
LGTM I'll help take some of the testing in a follow-up PR!

**Review by roeap — APPROVED** [review]
👍

**Pull request merged**